src/raftstore/store/store.rs

```diff
@@ -237,54 +254,74 @@ impl<T, C> Store<T, C> {
         // Scan region meta to get saved regions.
         let start_key = keys::REGION_META_MIN_KEY;
         let end_key = keys::REGION_META_MAX_KEY;
-        let engine = self.engine.clone();
+        let kv_engine = self.kv_engine.clone();
         let mut total_count = 0;
         let mut tomebstone_count = 0;
         let mut applying_count = 0;
 
         let t = Instant::now();
-        let mut wb = WriteBatch::new();
-        try!(engine.scan(start_key, end_key, false, &mut |key, value| {
-            let (region_id, suffix) = try!(keys::decode_region_meta_key(key));
-            if suffix != keys::REGION_STATE_SUFFIX {
-                return Ok(true);
-            }
+        let mut kv_wb = WriteBatch::new();
+        let mut raft_wb = WriteBatch::new();
+        try!(kv_engine.scan(
+            start_key,
+            end_key,
+            false,
+            &mut |key, value| {
+                let (region_id, suffix) = try!(keys::decode_region_meta_key(key));
+                if suffix != keys::REGION_STATE_SUFFIX {
+                    return Ok(true);
+                }
 
-            total_count += 1;
+                total_count += 1;
 
-            let local_state = try!(protobuf::parse_from_bytes::<RegionLocalState>(value));
-            let region = local_state.get_region();
-            if local_state.get_state() == PeerState::Tombstone {
-                tomebstone_count += 1;
-                debug!(
-                    "region {:?} is tombstone in store {}",
-                    region,
-                    self.store_id()
-                );
-                self.clear_stale_meta(&mut wb, region);
-                return Ok(true);
-            }
-            let mut peer = try!(Peer::create(self, region));
+                let local_state = try!(protobuf::parse_from_bytes::<RegionLocalState>(value));
+                let region = local_state.get_region();
+                if local_state.get_state() == PeerState::Tombstone {
+                    tomebstone_count += 1;
+                    debug!(
+                        "region {:?} is tombstone in store {}",
+                        region,
+                        self.store_id()
+                    );
+                    self.clear_stale_meta(&mut kv_wb, &mut raft_wb, region);
```

Comment:
clear stale meta will clear the region state key, do we still need to keep tombstone here?